1213 - Ability to add captions to video nodes

zifgu opened the pull request:
## Changes
As part of our accessibility goals, we would like to support adding captions to video nodes. This will apply to videos with format "mp4" (regular URL) and "kaltura", but not YouTube videos, since these already have caption support.

Kaltura's API includes captions, which are displayed in the Kaltura Media Player, so captions added to Kaltura videos should integrate with the Kaltura API. The interface for adding captions to regular URL videos should appear the same to users, but will not interface with the Kaltura API.

Changes to the video authoring form:
- Have a toggle “Add captions” that is off by default
- If the toggle button is on, display one row for captions
- The row should display:
    - A field for the caption label - default should be “CC” (but user could edit it)
    - A file upload field
    - A language field (English by default)
- If there is already one caption row that is filled in, then also show an “Add another caption” button
- If there are 2 or more rows, add a button similar to Kaltura to “Set as default” and whichever one is the default should have a text saying “Default” instead

No changes should be made on Kaltura until the video form is submitted.

## Screenshot
* This is a space to include a screenshot or GIF if this is a frontend change. Delete this section if this is not a frontend change.
## Issue Linkage
Closes #1213
## PR Dependency
Depends on: #1169 
## Automated Testing
* Please enter what automated testing you have added so that testers will know what they should NOT test

legendword commented on endpoints.php:
```diff
@@ -1743,8 +1764,14 @@ function perform_batched_upload_to_kaltura($videos, $use_kaltura_player)
                 if ($response->status === EntryStatus::READY) {
                     $node = save_video_upload_status($video, $videos_to_upload, UploadStatus::COMPLETE);
-                    TapestryHelpers::saveAndDeleteLocalVideo($node, $response, $use_kaltura_player, $video->file->file_path);
+                    TapestryHelpers::saveAndDeleteLocalVideo($node, $kalturaApi, $response, $use_kaltura_player, $video->file->file_path);
```
The kalturaApi parameter does not seem to be used in the implementation of TapestryHelpers::saveAndDeleteLocalVideo. Could you explain why this parameter was added?

zifgu replied:
Nice catch. This was a leftover piece of code and that argument shouldn't be there. I have just removed it.

legendword reviewed (changes requested):
Tested adding a caption to a URL Video node. The functionality works as expected, and the caption shows up while playing the video. The caption will show up by default, but the built-in media player in Chrome allows changing the caption track or turning it off. The default track is correctly set to the one specified in the NodeModal. Adding captions to a Kaltura video node works just as well. The Kaltura player correctly identifies the captions and allows changing a track or turning it off.

<img width="933" alt="image" src="https://user-images.githubusercontent.com/12668055/184694333-4db5d869-0780-40bb-993d-ace6f6b095d2.png">

However, for authoring Kaltura captions, I encountered some issues. When I added a couple caption tracks and then turned off the captions toggle (effectively clearing the caption tracks), the Kaltura player indeed indicates there are no available captions. Then, when I try to add a caption track to the same video, I encountered a 500 internal server error and got no feedback from the client (thus indicating it's not a TapestryError but rather a code-related error).

<img width="1439" alt="Pasted Graphic 4" src="https://user-images.githubusercontent.com/12668055/184696375-331092ef-0cda-4acf-89de-b3c8e00317f7.png">

Afterwards, when I play the video, a 404 error is printed in the console trying to locate a local vtt file, even though there shouldn't be any local caption files associated with the node at the moment.

<img width="304" alt="image" src="https://user-images.githubusercontent.com/12668055/184696449-882ddb0c-3e67-491b-ae7e-7d8c82ff426f.png">

When I open the NodeModal again, I see that the caption I was trying to add appeared with a URL that is on the Kaltura server. But this URL leads to an empty file, meaning that the caption file content was not successfully uploaded but the caption asset was created.

After I removed this broken caption by turning off the Captions toggle, the caption is successfully removed and I am no longer encountering the 404 error in the console whenever I play the video.

Hopefully the descriptions are enough for you to start debugging. Besides this issue, the feature looks really good. I left some refactoring comments and some confusions in the code, mainly on the new VideoForm component, that have to do with adhering to the coding conventions we have across the other form components.

zifgu commented:
> However, for authoring Kaltura captions, I encountered some issues. When I added a couple caption tracks and then turned off the captions toggle (effectively clearing the caption tracks), the Kaltura player indeed indicates there are no available captions. Then, when I try to add a caption track to the same video, I encountered a 500 internal server error and got no feedback from the client (thus indicating it's not a TapestryError but rather a code-related error).

Thanks for finding this bug. However, I wasn't able to reproduce it based on my understanding. I created a Kaltura video, added captions, saved it, removed the captions, saved it again, and tried to add captions again. The caption was added successfully the second time. When you have a moment, could you let me know the steps you followed?

zifgu commented on templates/vue/src/utils/Helpers.js:
```diff
@@ -147,8 +149,20 @@ export default class Helpers {
+          if (other.hasOwnProperty(prop)) {
+            return customCompareProps[prop](src[prop], other[prop])
+          }
+        }
```
This makes sense, thanks for catching it. I have just made these changes.

zifgu commented on templates/vue/cypress/integration/node-authoring.spec.js:
```diff
@@ -239,7 +239,7 @@ describe("Node Authoring", () => {
 
         cy.getByTestId(`node-video-url`)
-          .clear()
+          .clear({ force: true })
```
I had to add this because the node video URL field is covered by the "Operation failed" banner that was generated earlier in the test.

legendword replied:
Okay that sounds good!

zifgu commented on templates/vue/src/components/modals/common/FileUpload.vue:
```diff
@@ -1,5 +1,38 @@
+      :placeholder="placeholder"
+      :value="value"
+      :data-qa="inputTestId"
+      :data-testid="inputTestId"
```
Sounds good.

zifgu commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/CaptionRow.vue:
```diff
@@ -0,0 +1,180 @@
+          class="mr-1"
+          size="sm"
+          :variant="isRemovable ? 'danger' : 'light'"
+          :disabled="!isRemovable"
```
I think it was just a choice; when the user turns captions "On", they must provide at least one caption. If it would be clearer, I can change it so that deleting the last caption turns off the caption toggle.

legendword replied:
I think that makes more sense. Also since the toggle's state is related to the number of captions, it shouldn't require more code for it to work.

zifgu commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/VideoForm.vue:
```diff
@@ -18,68 +18,171 @@
   computed: {
-    mediaURL: {
+    mediaFormat: {
       get() {
-        return this.$store.state.currentEditingNode.typeData.mediaURL
+        return this.$store.state.currentEditingNode.mediaFormat
+      },
```
Fixed!

zifgu commented:
After testing with Kaltura's automatic captioning, some changes are required to account for Kaltura's automatic captioning service, which adds captions after a video is uploaded:
- Should accept media URLs of SRT format in the VideoForm caption editor.
- Add a "Display in Player" toggle to captions so that users can control the visibility of their captions and fix any captions Kaltura has automatically hidden.

zifgu commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/VideoForm.vue:
```diff
@@ -126,6 +268,70 @@ export default {
+      })
+    },
+    removeCaption(index, caption) {
+      this.captions = this.captions.filter((c, i) => i != index)
```
Fixed. Also made the same change when deleting pending captions.

zifgu commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/VideoForm.vue:
```diff
@@ -126,6 +268,70 @@ export default {
+          ...Helpers.deepCopy(defaultCaption),
+          id: Helpers.createUUID(),
+        },
+      ]
```
Made this change. I think this code is just left over from when I tried a different method where Array.push didn't work.

zifgu commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/VideoForm.vue:
```diff
@@ -18,68 +18,171 @@
+    CaptionRow,
   },
   data() {
+    const captions = this.$store.state.currentEditingNode.typeData.captions ?? []
```
Thanks for these suggestions -- I had some trouble making the captions array reactive, and this is a very clear solution.

I have implemented your suggestions here and from https://github.com/tapestry-tool/tapestry-wp/pull/1215#discussion_r945983543, and verified that everything is updating properly in the Vue devtools.

zifgu commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/VideoForm.vue:
```diff
@@ -18,68 +18,171 @@
+      kalturaId: this.$store.state.currentEditingNode.typeData.kalturaId,
+      editingKalturaId: false,
+      captions: captions,
+      useCaptions: captions.length > 0,
```
Fixed.

zifgu commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/VideoForm.vue:
```diff
@@ -18,68 +18,171 @@
+    const captions = this.$store.state.currentEditingNode.typeData.captions ?? []
     return {
       useKaltura: false,
+      kalturaId: this.$store.state.currentEditingNode.typeData.kalturaId,
```
I implemented it this way because the new UI allows users to edit the Kaltura ID separately before submitting it, so the Kaltura ID would only be stored into the Vuex state when "Submit" is clicked. If you have any suggestions to make this cleaner, please let me know!

legendword replied:
Thanks for explaining, I think it makes sense and let's keep it this way.

zifgu commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/VideoForm.vue:
```diff
@@ -18,68 +18,171 @@
+import client from "@/services/TapestryAPI"
 import Helpers from "@/utils/Helpers"
 import { mapMutations } from "vuex"
+import CaptionRow from "./CaptionRow"
```
Fixed!

zifgu commented on endpoints.php:
```diff
@@ -2081,3 +2119,38 @@ function getKalturaVideoUrl($request)
+
+function getKalturaAvailableLanguages($request)
+{
+    if (LOAD_KALTURA) {
```
Thanks for bringing this up. I was also wondering about this. I think we show the "Use Kaltura" field in the VideoForm regardless of whether Kaltura is enabled on the backend, which means it is possible for these endpoints to be called when LOAD_KALTURA is false. However, I'm not sure how to respond when the user tries to edit an existing Kaltura video while Kaltura is not enabled on the backend (a rare situation, but possible).

Let's discuss this tomorrow, since this change may be better suited for 1168.

zifgu replied:
Update from our discussion: This change is a good idea, but since we plan to keep Kaltura always enabled in production, I will create a separate PR off of 1168 to fix it.

zifgu commented:
Changes to be made to support TYDE workflow:
- Allow users to upload SRT files on Kaltura videos only.

legendword reviewed (approved):
Tried uploading a video to Kaltura through the Kaltura tab in the Settings modal, and waited for a few minutes for the automatic captioning to be processed. Then, I was able to see the automatically generated caption in the NodeModal of the uploaded Kaltura video. Submitting the NodeModal still worked as expected and no errors were thrown. Tried adding new caption tracks by uploading a local file, which were also working correctly. The custom labels on the caption tracks appeared in the Kaltura player. I was able to hide captions from the player by turning off the "Display in player" toggle. SRT caption files were successfully uploaded and the captions contained in the files were shown in the player.

As for the bug I mentioned in the review earlier, it was due to copying the local caption file URL to the FileUpload input directly. Since submitting the NodeModal on a Kaltura node uploads the local caption files and deletes them, that URL pointed to a nonexistent file. Uploading a new file every time fixed it. However, I would suggest returning an error when the given file URL for the caption does not exist. Currently if I input a nonexistent URL in the file input instead of clicking on the Upload button, submitting the NodeModal results in 500 server error and no error message is given to the user.

Other than that, everything looks good!

zifgu commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/VideoForm/CaptionRow.vue:
```diff
@@ -23,6 +23,14 @@
+        <b-form-checkbox
+          v-model="caption.displayOnPlayer"
+          switch
+          :data-qa="`caption-display-toggle-${index}`"
```
True... that makes a lot of sense. Fixed!

zifgu commented:
Great! Thank you for giving this PR such a detailed test.

> However, I would suggest returning an error when the given file URL for the caption does not exist. Currently if I input a nonexistent URL in the file input instead of clicking on the Upload button, submitting the NodeModal results in 500 server error and no error message is given to the user.

This makes a lot of sense since it's such a common case - having this error message would make it much easier to diagnose. Unfortunately, I think NodeModal does not currently display error messages to the user when an API call fails with an error. We might wish to fix this in the future.

For now, I can add a new TapestryError when the file does not exist. I will also make the corresponding change in 1168 for video uploads.

zifgu commented:
> Unfortunately, I think NodeModal does not currently display error messages to the user when an API call fails with an error. We might wish to fix this in the future.

Corrected - we may just need to catch errors and use `addApiError` to display them in the NodeModal. I will implement catching Kaltura-related errors in #1169 and here.

legendword reviewed (commented):
Tested giving an invalid URL for Kaltura caption upload and got feedback letting me know that the upload failed:

<img width="791" alt="image" src="https://user-images.githubusercontent.com/12668055/185693306-1dfe1fac-f91c-41c4-843a-a4f3374315df.png">

However, the "provided information" was not provided to the user. Even though I could see the server responded with the correct error message, it was not being displayed in the frontend because we didn't invoke the `addApiError` Vuex action. I would suggest looking at the other `addApiError` code we already have. I think what we usually do is instead of directly calling `client.*` in the component, we call a Vuex action that calls the `client.*` method, and this action will catch any errors and then call `dispatch("addApiError", error)`. So we should implement something like this for handling errors.

<img width="931" alt="image" src="https://user-images.githubusercontent.com/12668055/185693483-431c5f6f-22f1-4ddc-af97-debd5023f59d.png">

^ backend successfully returns the correct message, but frontend does not display this error message due to not using `addApiError`.

Another thing is I noticed you are no longer catching exceptions when adding captions to the Kaltura service and removed the FAILED_TO_SAVE_CAPTIONS tapestry error in the backend. Could you explain why this is no longer necessary?

Other than the above points, the code changes all look great to me.

zifgu commented:
> Even though I could see the server responded with the correct error message, it was not being displayed in the frontend because we didn't invoke the `addApiError` Vuex action. I would suggest looking at the other `addApiError` code we already have. I think what we usually do is instead of directly calling `client.*` in the component, we call a Vuex action that calls the `client.*` method, and this action will catch any errors and then call `dispatch("addApiError", error)`. So we should implement something like this for handling errors.

Thanks so much for letting me know how to make the error message show up. I have implemented it with `addApiError`. However, I didn't use a Vuex action because based on our existing usage of actions, it may be weird to use an action to modify `currentEditingNode` right before the node is submitted.

> Another thing is I noticed you are on longer catching exceptions when adding captions to the Kaltura service and removed the FAILED_TO_SAVE_CAPTIONS tapestry error in the backend. Could you explain why this is no longer necessary?

Right - I removed it because in my testing, the file not found error was always the cause of FAILED_TO_SAVE_CAPTIONS. Catching it would have replaced the file not found error with the more generic error message from FAILED_TO_SAVE_CAPTIONS. This does mean that fatal errors due to bugs or unexpected errors from Kaltura won't be caught, but I think (?) this is true of all our endpoints. Please let me know what you think, though.

legendword commented:
> However, I didn't use a Vuex action because based on our existing usage of actions, it may be weird to have an action specifically to modify currentEditingNode right before the node is submitted.

Could you explain why it may be weird? We are already modifying `currentEditingNode` in lots of places in the `handleSubmit` function, before the node is submitted. For example, the calls to `this.update` here basically modifies `currentEditingNode`:

https://github.com/tapestry-tool/tapestry-wp/blob/767c747bd471343fd60891a449d72aeef041331d/templates/vue/src/components/modals/NodeModal/index.vue#L838-L860

...so it is certainly okay to just call `commit("setCurrentEditingNodeProperty", ...)` from within a Vuex action, even if the action is performed right before submitting (since we have the `await`, the request will be completed before the node is submitted). Then again, I see your point that it is different from the other updates we are doing within `handleSubmit`. None of the updates so far used an action. It's merely a preference, and I agree to keep it this way and not introduce actions here.

Nevertheless, every time we are calling some method within the `client` util, we will be expecting an error. I would say we need to add try-catch to every call in the form of `client.*`, not just `client.updateKalturaVideoCaptions`. I'm noting it here because the calls are mostly added in this PR, but not adding try-catch does not result in any major consequences for this PR. The effects are very prominent in PR #1239 because the "Kaltura not available" error is thrown but yet again not being caught in the frontend. See the [review notes](https://github.com/tapestry-tool/tapestry-wp/pull/1239#pullrequestreview-1081133914) I left for #1239 for more details.

> This does mean that fatal errors due to bugs or unexpected errors from Kaltura won't be caught, but I think (?) this is true of all our endpoints.

Unexpected errors resulting from interacting with the Kaltura are already hard to detect and debug since there is no reliable way to trigger them, so I think it should be fine. Thanks for clearing that up for me!

zifgu commented:
> Nevertheless, every time we are calling some method within the `client` util, we will be expecting an error. I would say we need to add try-catch to every call in the form of `client.*`, not just `client.updateKalturaVideoCaptions`. I'm noting it here because the calls are mostly added in this PR, but not adding try-catch does not result in any major consequences for this PR. The effects are very prominent in PR #1239 because the "Kaltura not available" error is thrown but yet again not being caught in the frontend.

That makes sense. I thought #1239 would be a more suitable place for the fix, since it is only in #1239 that the Kaltura endpoints start to throw errors and the "Kaltura not available" error is added, so I have added try-catch to the `client.*` calls there.

zifgu commented:
Changes since last review that may need testing:
- Adding a caption with an error: error message should be displayed on the frontend and the user should not be able to submit.
- Uploading a video where the caption has an error (e.g. file does not exist): upload should complete without errors, but the Additional Info should show that "x captions failed to upload" and the failed captions should appear in the pending captions.

legendword reviewed (commented):
Tested adding a nonexistent caption file to a URL video. Submitting the node modal resulted in an error message saying the file could not be found, which is the correct behaviour.

Tested adding one valid caption file and one nonexistent caption file to a URL video, then uploading the video node to Kaltura using the Kaltura modal with "Switch to Kaltura player" toggled on. The video was uploaded without errors, but with a message in the additional info section describing the failed caption upload:

<img width="779" alt="image" src="https://user-images.githubusercontent.com/12668055/186794481-59407710-7616-4a0c-a10f-fa2d766899a9.png">

Without refreshing the page, I opened the node modal for that node, and everything was what they were before the upload, which is expected since the user would need to refresh to see the updated node information. After a refresh, opening the node modal shows two captions and one pending caption. Of the two captions, one of them was the one that failed to upload. It still had the other settings (language, customized label) but had a different source URL that ended in `?.`, basically having no file extension. The pending caption had the exact same settings as the failed-to-upload caption (the source URL was the local nonexistent URL). I just want to confirm that this is indeed the behaviour I should be expecting. At this moment, I can see two caption tracks when playing the Kaltura video. I'm assuming that the failed caption should just be in the pending captions and not in the captions list, and it should not show up in the Kaltura player.

<img width="778" alt="image" src="https://user-images.githubusercontent.com/12668055/186795459-32909f13-3e4d-49c4-8962-9ae5351fe6b9.png">

^ Failed caption, but in the captions list with a URL that has no extension.

<img width="773" alt="image" src="https://user-images.githubusercontent.com/12668055/186795517-f176c5ef-c479-4c48-a9c2-c8219bbc25f3.png">

^ Pending caption, working as expected.

Also reviewed the code changes and they look good. I had one question while viewing the code changes that requires some explanation.

zifgu commented:
> I just want to confirm that this is indeed the behaviour I should be expecting. At this moment, I can see two caption tracks when playing the Kaltura video. I'm assuming that the failed caption should just be in the pending captions and not in the captions list, and it should not show up in the Kaltura player.

Unfortunately, this is a known issue. The series of requests to create a caption is "create caption", "upload file", "set caption content to the file data", but to achieve acceptable performance, I am submitting all the requests to Kaltura in one HTTP request via Kaltura's Multi-Request. This means that if the "upload file" step fails, the "create caption" step will still be processed. Without content, the created caption will be in an "unfinished" state (on the Kaltura admin dashboard, you will see that its status is "Processing"), but Kaltura seems to still show these in the media player by default. I am currently returning "Processing" captions to the user so that they have the chance to edit these captions and potentially fix whatever is wrong. This is why you'll see the caption with the error in the list.

This will be resolved by adding the pending caption back to the list -- pending captions are merged by ID, so it will replace the "Processing" caption with new content that is hopefully valid. Unfortunately as you've seen, it is a bit different from the expected behaviour.

I think I made the attempt before, but I can have another look at this and try to fix it.

zifgu commented on classes/class.kaltura-api.php:
```diff
@@ -336,9 +336,9 @@ private function _processResponses($kclient, $responses, $toAdd, $toUpdate)
+                    array_push($pending, $this->_filterCaptionAsset($kclient, $caption, $caption->captionUrl, null, $metadataErrorMessage));
                 } elseif ($contentError) {
-                    array_push($pending, $this->_filterCaptionAsset($kclient, $caption, $caption->captionUrl, $contentErrorMessage));
+                    array_push($pending, $this->_filterCaptionAsset($kclient, $caption, $caption->captionUrl, $metadataResponse->id, $contentErrorMessage));
```
In this case, there is no `metadataError`, which means the caption was successfully created on Kaltura, has a Kaltura ID, and will be displayed in the list of captions in the NodeModal. However, `$caption` has the original ID generated for this caption on the frontend, so in order for the frontend to match this pending caption with the caption on Kaltura, we need to overwrite its id with the Kaltura ID of the newly added caption.

legendword replied:
Okay, that makes more sense now. Thanks for explaining!

zifgu commented:
> the failed caption should just be in the pending captions and not in the captions list, and it should not show up in the Kaltura player.

Just implemented a fix that hopefully makes this the case.

legendword reviewed (approved):
The code changes look great. Now uploading a nonexistent caption file will only result in a pending caption.

When I try to add the pending caption back and submit the modal without changing anything (so the caption file is still nonexistent), the submission failed as expected saying that the file was not found on the server. However, this time the failed caption is not added to pending captions but instead removed from the captions altogether. I think this is the expected behaviour from what I can see in the code. Just wanted to point this out in case this behaviour wasn't known before.

zifgu commented:
> However, this time the failed caption is not added to pending captions but instead removed from the captions altogether.

Wait -- I think this shouldn't happen in general. When I just tested the scenario locally, everything in the VideoForm is preserved after submitting, including the failed caption, since the file not found error interrupts submission. The failed caption should only disappear if you edit and re-submit the Kaltura ID to force a reload -- it is expected in this case, because the reload overwrites the captions list, and the failed caption would not be among the captions on Kaltura.

When you have a moment, can you share what happened with a bit more detail? I'd just like to confirm that this isn't a bug.

zifgu commented:
> Sure! Here is a recording of the process:
> 
>  [Screen.Recording.2022-08-29.at.10.15.39.mp4
](https://user-images.githubusercontent.com/12668055/187260816-1b80152b-eb20-4f10-81a1-db77da5fccb1.mp4)

Thanks! I believe I see what's going on -- the failed video being overwritten is expected, but the unexpected situation is due to a bug. After catching the error, the subsequent code that updates the node data was still running, which is also why there are no more unsaved changes. I've fixed this so that we return from `handleSubmit` upon catching an error.

Bitajkr commented:
I am unable to upload a .vtt caption file as I see the following error: 
<img width="871" alt="Screen Shot 2022-10-24 at 2 36 39 PM" src="https://user-images.githubusercontent.com/113054791/197634516-21da2bec-a022-40a1-b195-ed7f4e3c4787.png">

wynnset commented:
Hi @Bitajkr I've put in a fix for this. Can you try and let me know if it works now, please?

Bitajkr commented:
> Hi @Bitajkr I've put in a fix for this. Can you try and let me know if it works now, please?

Hi @wynnset Thanks for fixing this! It's working now. I will continue to test this feature and will update you by the end of the week.

Bitajkr commented:
Here are some things that I have noticed so far that do not behave as I had expected: 

- When I try to add a Youtube video, if I do not add the link to the youtube video, the page reverts to the "file upload" tab instead of staying on the Youtube tab (see video below). 
[Screen Recording 2022-10-26 at 1.29.45 PM.zip](https://github.com/tapestry-tool/tapestry-wp/files/9873535/Screen.Recording.2022-10-26.at.1.29.45.PM.zip)
-  I am unable to add a link to a Vimeo video under the "Video" content type (e.g., https://vimeo.com/ondemand/witchcupcakes)
- I closed the browser right after pressing "Save and Publish" on a node with a video + captions and when I revisited the tapestry, the node was there. I think that this is the expected behaviour, but I thought I should point this out. 
- I tried uploading an MP4 URL to Tapestry Tool and it did not work as the video node would not open (see video below).
[Screen Recording 2022-10-26 at 1.51.23 PM.zip](https://github.com/tapestry-tool/tapestry-wp/files/9873679/Screen.Recording.2022-10-26.at.1.51.23.PM.zip)

wynnset commented:
> * I tried uploading an MP4 URL to Tapestry Tool and it did not work as the video node would not open

@Bitajkr can you give me the URL of the video you were trying to add? Also, the alert message is cut off in your video, but I can recreate it on my end. (by the way, you can always attach videos directly in your comment so they are viewable on github, rather than zipping them)

wynnset commented:
> * I am unable to add a link to a Vimeo video under the "Video" content type (e.g., https://vimeo.com/ondemand/witchcupcakes)

We have never supported Vimeo and still do not.

wynnset commented:
> * I closed the browser right after pressing "Save and Publish" on a node with a video + captions and when I revisited the tapestry, the node was there. I think that this is the expected behaviour, but I thought I should point this out.

I'm sorry I don't think I understand what this means. Why would the node not be there?

Bitajkr commented:
> > * I am unable to add a link to a Vimeo video under the "Video" content type (e.g., https://vimeo.com/ondemand/witchcupcakes)
> 
> We have never supported Vimeo and still do not.

Sounds good, just wanted to test this out as a URL to see if it would work.

Bitajkr commented:
> 

Here's the MP4 URL: https://tapestry-tool.com/videofiles/promo_lowres.mp4

Also, GitHub wasn't allowing me to directly attach the videos as their file size was too big. So instead, I attached them as Zip files.

Bitajkr commented:
I cannot see the "Kaltura" tab in the Tapestry Settings (see image). Is this tab supposed to always be visible or does it only become visible after uploading a certain type of video. 

<img width="977" alt="Screen Shot 2022-10-26 at 3 24 05 PM" src="https://user-images.githubusercontent.com/113054791/198149985-2fc53cd2-9ae8-4a79-bb3d-e0706258aa4b.png">

wynnset commented:
> I cannot see the "Kaltura" tab in the Tapestry Settings (see image)

Sorry Bita, I had to enable Kaltura for this site. You can access the Kaltura settings by clicking on the "..." icon next to the depth slider and selecting "Upload to Kaltura"
<img width="521" alt="image" src="https://user-images.githubusercontent.com/8981258/198153477-5675350c-3724-4a8f-b7c8-d3a338bd0e23.png">

Please see #1169 for more information about the basic Kaltura functionality that we've implemented.

Bitajkr commented:
> > I cannot see the "Kaltura" tab in the Tapestry Settings (see image)
> 
> Sorry @Bitajkr , I had to enable Kaltura for this site. You can access the Kaltura settings by clicking on the "..." icon next to the depth slider and selecting "Upload to Kaltura" <img alt="image" width="521" src="https://user-images.githubusercontent.com/8981258/198153477-5675350c-3724-4a8f-b7c8-d3a338bd0e23.png">
> 
> Please see #1169 for more information about the basic Kaltura functionality that we've implemented.

Thanks @wynnset for the clarification; however, unfortunately, I still can't see the "upload to Kaltura" button (see image). 
<img width="613" alt="Screen Shot 2022-10-26 at 3 56 46 PM" src="https://user-images.githubusercontent.com/113054791/198154090-34fd7191-0e4b-4571-b654-ccec37d79064.png">

wynnset commented:
@Bitajkr can you clarify which site you are testing on? A link to that tapestry would be great.

wynnset commented:
> Here's the MP4 URL: https://tapestry-tool.com/videofiles/promo_lowres.mp4

It seems like the problem here is with the video. I guess Wix doesn't allow other sites to link to media on its site. That's why you cannot see the video. This is probably the same case if you try to do the same thing on the existing production code. So basically this is not going to get fixed. If we want to include a more friendly message, we could do it, but we should open a different issue for that on Github since it's not related to this feature.

Bitajkr commented:
While trying to test this feature out, I tried uploading some of the Deaf Module videos from BC SUPPORT and noticed that none of these particular videos could be uploaded. I am not entirely sure what was causing this. Perhaps, could this be a result of a restriction on size of uploads?

Bitajkr commented:
Aside from my previous message, this feature seems to be working well, such that I am easily able to upload captions to a video and have them shown in the lightbox.

legendword reviewed (approved):
Because it has been a long time since the last review, I did some testing covering most of the use cases of the features in this pull request as well as did some code fixes and refactoring. Below are all the things I tested:

- [x]  Changing the language and label of an existing caption on a Kaltura video. The fields are updated on the KMC, but the changes are not reflected immediately in the player when playing the video, possibly due to caching. After a few minutes, the changes can be seen from the player.
- [x]  Changing fields related to captions on the KMC and seeing the changes be reflected in the node modal when editing the Kaltura video. The user will be prompted to save the node since the captions data has changed.
- [x]  Uploading a .vtt caption file to a Kaltura video and saving the node. Again, the caption shows up immediately on KMC, but the caption does not show up immediately in the player. Re-opening the node modal for editing will show “You have unsaved changes” due to a difference from the captions data saved in the node and the captions data fetched from Kaltura. Ideally the two shouldn’t be different, but in this case the different seems to be only in the order of the captions so it does not affect normal usage.
- [x]  Putting in an invalid local URL in the caption URL field results in error stating “Uploaded file was not found on the server” and prevents user from saving the node.
- [x]  Using the regular HTML5 player to play Kaltura videos and seeing the captions associated with that Kaltura video.
- [x]  Uploading a video file to Kaltura using the node modal and seeing the Kaltura ID after successful upload. Subsequent additions to the captions list also works.
- [x]  Deleting the only caption in the captions list for a Kaltura video.
- [x]  Adding and editing captions for a non-Kaltura URL video. All options related to each caption works as expected.
- [x]  Uploading an existing local video node that contains two valid captions to Kaltura via the upload dashboard. Completes successfully with all caption data.
- [x]  Uploading an existing local video node that contains one valid and one invalid caption to Kaltura via the upload dashboard. Completes successfully with information displaying 1 of the captions failed to upload. After refreshing the page and opening the node modal, the failed caption is only in the Pending Captions list.

Overall, everything is working well and the features are complete with no apparent bugs. The delay of a few minutes for the Kaltura player to reflect the caption changes was something not discovered or discussed before, so it might be a new behaviour possibly due to some differences in the player settings when we migrated to the new Kaltura platform.

wynnset commented on endpoints/endpoints.kaltura.php:
```diff
@@ -291,19 +312,6 @@ public static function getKalturaUploadStatus($request)
         ];
     }
 
-    /**
-     * Save Kaltura upload status
-     */
-    public static function saveVideoUploadStatus($video, $videosToUpload, $newStatus, $kalturaData = null)
-    {
-        $video->uploadStatus = $newStatus;
-        self::_updateUploadLog($videosToUpload);
-
-        $node = new TapestryNode($video->tapestryID, $video->nodeID);
-        self::_saveVideoUploadStatusInNode($node, $newStatus, $kalturaData);
-        return $node;
-    }
-
```
@legendword can you please confirm that this was removed intentionally? Sorry I merged this in, but then noticed this right after.
